linkerd/proxy/http/src/client.rs

```diff
@@ -114,48 +75,59 @@ where
 
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.connect.poll_ready(cx).map_err(Into::into)
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, target: T) -> Self::Future {
         trace!("Building HTTP client");
         let connect = self.connect.clone();
-        match target.http_settings() {
-            Settings::Http1 {
-                keep_alive,
-                wants_h1_upgrade: _,
-                was_absolute_form,
-            } => {
-                let mut h1 = hyper::Client::builder();
-                let h1 = if !keep_alive {
-                    // disable hyper's connection pooling by setting the maximum
-                    // number of idle connections to 0.
-                    h1.pool_max_idle_per_host(0)
-                } else {
-                    &mut h1
+        let h2_settings = self.h2_settings.clone();
+
+        Box::pin(async move {
+            match *target.as_ref() {
+                Settings::Http1 {
+                    keep_alive,
+                    wants_h1_upgrade: _,
+                    was_absolute_form,
+                } => {
+                    let mut h1 = hyper::Client::builder();
+                    let h1 = if !keep_alive {
+                        // disable hyper's connection pooling by setting the maximum
+                        // number of idle connections to 0.
+                        h1.pool_max_idle_per_host(0)
+                    } else {
+                        &mut h1
+                    }
+                    // hyper should only try to automatically
+                    // set the host if the request was in absolute_form
+                    .executor(trace::Executor::new())
+                    .set_host(was_absolute_form)
+                    .build(HyperConnect::new(
+                        connect,
+                        target,
+                        was_absolute_form,
+                    ));
+                    Ok(Client::Http1(h1))
+                }
+                Settings::Http2 => {
+                    let h2 = h2::Connect::new(connect, h2_settings)
+                        .oneshot(target)
+                        .await?;
+                    Ok(Client::Http2(h2))
                 }
-                // hyper should only try to automatically
-                // set the host if the request was in absolute_form
-                .executor(trace::Executor::new())
-                .set_host(was_absolute_form)
-                .build(HyperConnect::new(connect, target, was_absolute_form));
-                MakeFuture::Http1(Some(h1))
-            }
-            Settings::Http2 => {
-                let h2 = h2::Connect::new(connect, self.h2_settings.clone()).oneshot(target);
-                MakeFuture::Http2(h2)
             }
-        }
+        })
```

Comment:
it seems like we _could_ do this one pretty easily with `futures::Either` and like, one combinator, since there's only a single `await` in the h2 arm, and no awaits in the h1 arm...may not be worth it, the async fn version is still slightly nicer.

Reply:
Assume, for now, that this is going to change and become more complicated